File: lib/ai.ts
const OPENROUTER_API_KEY = process.env.OPENROUTER_API_KEY;
const OPENROUTER_BASE_URL = process.env.OPENROUTER_BASE_URL || "https://openrouter.ai/api/v1";

// Free models with fallback order
const MODELS = [
  "meta-llama/llama-3.1-8b-instruct",
  "mistralai/mistral-7b-instruct",
  "google/gemma-7b",
  "qwen/qwen-2.5-7b-instruct",
];

export async function callOpenRouter(
  prompt: string,
  systemPrompt: string = "You are an educational AI assistant.",
  modelIndex: number = 0
): Promise<string> {
  if (!OPENROUTER_API_KEY) {
    throw new Error("OPENROUTER_API_KEY is not set");
  }

  const model = MODELS[modelIndex];

  try {
    const response = await fetch(`${OPENROUTER_BASE_URL}/chat/completions`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "Authorization": `Bearer ${OPENROUTER_API_KEY}`,
      },
      body: JSON.stringify({
        model,
        messages: [
          { role: "system", content: systemPrompt },
          { role: "user", content: prompt },
        ],
      }),
    });

    if (!response.ok) {
      if (modelIndex < MODELS.length - 1) {
        // Try next model
        return callOpenRouter(prompt, systemPrompt, modelIndex + 1);
      }
      throw new Error(`OpenRouter API error: ${response.statusText}`);
    }

    const data = await response.json();
    return data.choices[0]?.message?.content || "No response from AI";
  } catch (error) {
    if (modelIndex < MODELS.length - 1) {
      // Try next model
      return callOpenRouter(prompt, systemPrompt, modelIndex + 1);
    }
    throw error;
  }
}

// 1. Paper Generation
export async function generateExamPaper(
  subject: string,
  classLevel: string,
  examType: "mcq" | "short" | "long" | "full",
  topics: string[],
  totalMarks: number = 100
): Promise<{
  questions: Array<{
    type: string;
    question: string;
    marks: number;
    options?: string[];
    correctAnswer?: string;
  }>;
  answerKey: string;
}> {
  const prompt = `Generate a ${examType} exam paper for ${subject} for class ${classLevel}.
Topics to cover: ${topics.join(", ")}
Total marks: ${totalMarks}

Requirements:
${examType === "mcq" ? "- Generate 20 multiple choice questions with 4 options each" : ""}
${examType === "short" ? "- Generate 10 short answer questions (2-3 marks each)" : ""}
${examType === "long" ? "- Generate 5 long answer questions (10-15 marks each)" : ""}
${examType === "full" ? "- Generate a complete exam paper with MCQs, short questions, and long questions" : ""}

Format the response as JSON with this structure:
{
  "questions": [
    {
      "type": "mcq|short|long",
      "question": "question text",
      "marks": number,
      "options": ["option1", "option2", ...] (only for MCQ),
      "correctAnswer": "answer" (only for MCQ)
    }
  ],
  "answerKey": "detailed answer key"
}`;

  const response = await callOpenRouter(
    prompt,
    "You are an expert educational content generator. Generate high-quality exam papers."
  );

  try {
    const jsonMatch = response.match(/\{[\s\S]*\}/);
    if (jsonMatch) {
      return JSON.parse(jsonMatch[0]);
    }
  } catch (e) {
    // Fallback if JSON parsing fails
  }

  return {
    questions: [],
    answerKey: response,
  };
}

// 2. AI Auto-Grading
export async function gradeStudentAnswer(
  studentAnswer: string,
  correctAnswer: string,
  question: string,
  maxMarks: number
): Promise<{
  marks: number;
  explanation: string;
  feedback: string;
}> {
  const prompt = `Grade this student answer:

Question: ${question}
Correct Answer: ${correctAnswer}
Student Answer: ${studentAnswer}
Maximum Marks: ${maxMarks}

Provide:
1. Marks awarded (out of ${maxMarks})
2. Detailed explanation of grading
3. Constructive feedback for the student

Format as JSON:
{
  "marks": number,
  "explanation": "why these marks were awarded",
  "feedback": "feedback for student"
}`;

  const response = await callOpenRouter(
    prompt,
    "You are an expert teacher grading student answers fairly and providing constructive feedback."
  );

  try {
    const jsonMatch = response.match(/\{[\s\S]*\}/);
    if (jsonMatch) {
      const parsed = JSON.parse(jsonMatch[0]);
      return {
        marks: Math.min(maxMarks, Math.max(0, parsed.marks || maxMarks * 0.5)),
        explanation: parsed.explanation || "AI grading completed",
        feedback: parsed.feedback || "Good effort!",
      };
    }
  } catch (e) {
    // Fallback
  }

  return {
    marks: maxMarks * 0.5,
    explanation: "AI grading completed",
    feedback: response,
  };
}

// 3. Student Performance Forecast
export async function predictStudentPerformance(
  studentData: {
    pastGrades: Array<{ subject: string; marks: number; totalMarks: number; date: string }>;
    attendance: { present: number; total: number; percentage: number };
    syllabusProgress: { completed: number; total: number; percentage: number };
  }
): Promise<{
  predictedGrade: string;
  confidence: string;
  factors: string[];
  recommendations: string[];
}> {
  const prompt = `Analyze this student's performance data and predict future performance:

Past Grades:
${studentData.pastGrades.map((g) => `${g.subject}: ${g.marks}/${g.totalMarks} (${((g.marks / g.totalMarks) * 100).toFixed(1)}%)`).join("\n")}

Attendance: ${studentData.attendance.present}/${studentData.attendance.total} (${studentData.attendance.percentage}%)

Syllabus Progress: ${studentData.syllabusProgress.completed}/${studentData.syllabusProgress.total} (${studentData.syllabusProgress.percentage}%)

Provide:
1. Predicted overall grade (A+, A, B+, B, C+, C, D, F)
2. Confidence level (High/Medium/Low)
3. Key factors affecting performance
4. Recommendations for improvement

Format as JSON:
{
  "predictedGrade": "A|B+|B|C+|C|D|F",
  "confidence": "High|Medium|Low",
  "factors": ["factor1", "factor2"],
  "recommendations": ["recommendation1", "recommendation2"]
}`;

  const response = await callOpenRouter(
    prompt,
    "You are an expert educational analyst predicting student performance based on data."
  );

  try {
    const jsonMatch = response.match(/\{[\s\S]*\}/);
    if (jsonMatch) {
      return JSON.parse(jsonMatch[0]);
    }
  } catch (e) {
    // Fallback
  }

  return {
    predictedGrade: "B",
    confidence: "Medium",
    factors: ["Based on current performance trends"],
    recommendations: ["Continue current study habits", "Improve attendance"],
  };
}

// 4. Teacher Insights
export async function generateTeacherInsights(
  teacherData: {
    subject: string;
    classAverage: number;
    studentCount: number;
    weakStudents: number;
    attendanceRate: number;
  }
): Promise<{
  analysis: string;
  strengths: string[];
  weaknesses: string[];
  recommendations: string[];
}> {
  const prompt = `Analyze this teacher's performance:

Subject: ${teacherData.subject}
Class Average: ${teacherData.classAverage}%
Number of Students: ${teacherData.studentCount}
Weak Students (below 50%): ${teacherData.weakStudents}
Attendance Rate: ${teacherData.attendanceRate}%

Provide:
1. Overall analysis
2. Strengths
3. Areas for improvement
4. Specific recommendations

Format as JSON:
{
  "analysis": "overall analysis text",
  "strengths": ["strength1", "strength2"],
  "weaknesses": ["weakness1", "weakness2"],
  "recommendations": ["recommendation1", "recommendation2"]
}`;

  const response = await callOpenRouter(
    prompt,
    "You are an expert educational consultant analyzing teacher performance."
  );

  try {
    const jsonMatch = response.match(/\{[\s\S]*\}/);
    if (jsonMatch) {
      return JSON.parse(jsonMatch[0]);
    }
  } catch (e) {
    // Fallback
  }

  return {
    analysis: "Performance analysis completed",
    strengths: ["Good class management"],
    weaknesses: ["Some students need extra attention"],
    recommendations: ["Provide additional support to weak students"],
  };
}

// 5. Homework Questions
export async function generateHomeworkQuestions(
  subject: string,
  topic: string,
  difficulty: "easy" | "medium" | "hard",
  count: number = 5
): Promise<{
  questions: Array<{
    question: string;
    type: string;
    difficulty: string;
  }>;
}> {
  const prompt = `Generate ${count} ${difficulty} homework questions for ${subject} on the topic: ${topic}

Format as JSON:
{
  "questions": [
    {
      "question": "question text",
      "type": "short|long|mcq",
      "difficulty": "easy|medium|hard"
    }
  ]
}`;

  const response = await callOpenRouter(
    prompt,
    "You are an expert teacher creating homework assignments."
  );

  try {
    const jsonMatch = response.match(/\{[\s\S]*\}/);
    if (jsonMatch) {
      return JSON.parse(jsonMatch[0]);
    }
  } catch (e) {
    // Fallback
  }

  return {
    questions: [],
  };
}

// 6. Detect Weak Students
export async function detectWeakStudents(
  studentsData: Array<{
    name: string;
    averageGrade: number;
    attendance: number;
    subjects: Array<{ name: string; grade: number }>;
  }>
): Promise<{
  weakStudents: Array<{
    name: string;
    reasons: string[];
    recommendations: string[];
  }>;
}> {
  const prompt = `Identify weak students from this data:

${studentsData.map((s) => `
Student: ${s.name}
Average Grade: ${s.averageGrade}%
Attendance: ${s.attendance}%
Subjects: ${s.subjects.map((sub) => `${sub.name}: ${sub.grade}%`).join(", ")}
`).join("\n")}

For each weak student (average < 50% or attendance < 75%), provide:
1. Reasons for poor performance
2. Specific recommendations

Format as JSON:
{
  "weakStudents": [
    {
      "name": "student name",
      "reasons": ["reason1", "reason2"],
      "recommendations": ["recommendation1", "recommendation2"]
    }
  ]
}`;

  const response = await callOpenRouter(
    prompt,
    "You are an expert educational counselor identifying and helping weak students."
  );

  try {
    const jsonMatch = response.match(/\{[\s\S]*\}/);
    if (jsonMatch) {
      return JSON.parse(jsonMatch[0]);
    }
  } catch (e) {
    // Fallback
  }

  return {
    weakStudents: [],
  };
}

// 7. Detect Syllabus Delay
export async function detectSyllabusDelay(
  syllabusData: {
    totalTopics: number;
    completedTopics: number;
    currentDate: string;
    expectedCompletionDate: string;
    subjects: Array<{ name: string; progress: number }>;
  }
): Promise<{
  isDelayed: boolean;
  delayPercentage: number;
  affectedSubjects: string[];
  recommendations: string[];
}> {
  const prompt = `Analyze syllabus progress:

Total Topics: ${syllabusData.totalTopics}
Completed: ${syllabusData.completedTopics}
Current Date: ${syllabusData.currentDate}
Expected Completion: ${syllabusData.expectedCompletionDate}

Subject Progress:
${syllabusData.subjects.map((s) => `${s.name}: ${s.progress}%`).join("\n")}

Determine:
1. Is the syllabus delayed?
2. Delay percentage
3. Most affected subjects
4. Recommendations to catch up

Format as JSON:
{
  "isDelayed": true|false,
  "delayPercentage": number,
  "affectedSubjects": ["subject1", "subject2"],
  "recommendations": ["recommendation1", "recommendation2"]
}`;

  const response = await callOpenRouter(
    prompt,
    "You are an expert educational planner analyzing syllabus progress."
  );

  try {
    const jsonMatch = response.match(/\{[\s\S]*\}/);
    if (jsonMatch) {
      return JSON.parse(jsonMatch[0]);
    }
  } catch (e) {
    // Fallback
  }

  const progress = (syllabusData.completedTopics / syllabusData.totalTopics) * 100;
  return {
    isDelayed: progress < 70,
    delayPercentage: Math.max(0, 100 - progress),
    affectedSubjects: syllabusData.subjects.filter((s) => s.progress < 70).map((s) => s.name),
    recommendations: ["Increase teaching hours", "Prioritize important topics"],
  };
}

File: backend/app/routers/ai.py
"""
AI endpoints
"""

from fastapi import APIRouter, HTTPException, status, Depends
from pydantic import BaseModel
from typing import List, Optional, Dict, Any
from app.core.auth import require_teacher, require_admin_or_principal, CurrentUser
from app.services.ai_service import (
    generate_homework_questions,
    generate_exam_paper,
    grade_student_answer,
    predict_student_performance,
    generate_teacher_insights
)
import logging

logger = logging.getLogger(__name__)

router = APIRouter()


# ============ Homework Generation ============

class GenerateHomeworkRequest(BaseModel):
    subject: str
    topic: str
    difficulty: str
    count: int


@router.post("/homework")
async def generate_homework(
    request: GenerateHomeworkRequest,
    user: CurrentUser = Depends(require_teacher)
):
    """Generate homework questions using AI"""
    try:
        questions = await generate_homework_questions(
            request.subject,
            request.topic,
            request.difficulty,
            request.count
        )
        
        return questions
    
    except Exception as e:
        logger.error(f"Error generating homework: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to generate homework: {str(e)}"
        )


# ============ Exam Paper Generation ============

class GeneratePaperRequest(BaseModel):
    subject: str
    classLevel: str
    examType: str
    topics: List[str]
    totalMarks: int = 100


@router.post("/generate-paper")
async def generate_paper(
    request: GeneratePaperRequest,
    user: CurrentUser = Depends(require_teacher)
):
    """Generate exam paper using AI"""
    try:
        paper = await generate_exam_paper(
            request.subject,
            request.classLevel,
            request.examType,
            request.topics,
            request.totalMarks
        )
        
        return paper
    
    except Exception as e:
        logger.error(f"Error generating paper: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to generate paper: {str(e)}"
        )


# ============ AI Grading ============

class GradeAnswerRequest(BaseModel):
    studentAnswer: str
    correctAnswer: str
    question: str
    maxMarks: int


@router.post("/grade")
async def grade_answer(
    request: GradeAnswerRequest,
    user: CurrentUser = Depends(require_teacher)
):
    """Grade student answer using AI"""
    try:
        result = await grade_student_answer(
            request.studentAnswer,
            request.correctAnswer,
            request.question,
            request.maxMarks
        )
        
        return result
    
    except Exception as e:
        logger.error(f"Error grading answer: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to grade answer: {str(e)}"
        )


# ============ Student Forecast ============

class ForecastRequest(BaseModel):
    pastGrades: List[Dict[str, Any]]
    attendance: Dict[str, Any]
    syllabusProgress: Dict[str, Any]


@router.post("/forecast")
async def predict_performance(
    request: ForecastRequest,
    user: CurrentUser = Depends(require_teacher)
):
    """Predict student performance"""
    try:
        student_data = {
            "pastGrades": request.pastGrades,
            "attendance": request.attendance,
            "syllabusProgress": request.syllabusProgress
        }
        
        result = await predict_student_performance(student_data)
        
        return result
    
    except Exception as e:
        logger.error(f"Error predicting performance: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to predict performance: {str(e)}"
        )


# ============ Teacher Insights ============

class InsightsRequest(BaseModel):
    type: str
    data: Dict[str, Any]


@router.post("/insights")
async def generate_insights(
    request: InsightsRequest,
    user: CurrentUser = Depends(require_admin_or_principal)
):
    """Generate AI insights"""
    try:
        if request.type == "teacher":
            result = await generate_teacher_insights(request.data)
        else:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Invalid insight type: {request.type}"
            )
        
        return result
    
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error generating insights: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to generate insights: {str(e)}"
        )

